plugins/node/instrumentation-runtime-node/src/metrics/heapSpacesSizeAndUsedCollector.ts

```diff
@@ -0,0 +1,132 @@
+    [V8HeapSpaceMetrics.heapLimit]: {
+      description: 'Total heap memory size pre-allocated.',
+    },
+    [V8HeapSpaceMetrics.used]: {
+      description: 'Heap Memory size allocated.',
+    },
+    [V8HeapSpaceMetrics.available]: {
+      description: 'Heap space available size.',
+    },
+    [V8HeapSpaceMetrics.physical]: {
+      description: 'Committed size of a heap space.',
+    },
+  };
+
+export class HeapSpacesSizeAndUsedCollector extends BaseCollector<
+  HeapSpaceInfo[]
+> {
+  constructor(
+    config: RuntimeNodeInstrumentationConfig = {},
+    namePrefix: string
+  ) {
+    super(config, namePrefix);
+  }
+
+  updateMetricInstruments(meter: Meter): void {
+    const heapLimit = meter.createObservableGauge(
+      `${this.namePrefix}.${V8HeapSpaceMetrics.heapLimit}`,
+      {
+        description: metricNames[V8HeapSpaceMetrics.heapLimit].description,
+        unit: 'By',
+      }
+    );
+    const heapSpaceUsed = meter.createObservableGauge(
+      `${this.namePrefix}.${V8HeapSpaceMetrics.used}`,
+      {
+        description: metricNames[V8HeapSpaceMetrics.used].description,
+        unit: 'By',
+      }
+    );
+    const heapSpaceAvailable = meter.createObservableGauge(
+      `${this.namePrefix}.${V8HeapSpaceMetrics.available}`,
+      {
+        description: metricNames[V8HeapSpaceMetrics.available].description,
+        unit: 'By',
+      }
+    );
+    const heapSpacePhysical = meter.createObservableGauge(
+      `${this.namePrefix}.${V8HeapSpaceMetrics.physical}`,
+      {
+        description: metricNames[V8HeapSpaceMetrics.physical].description,
+        unit: 'By',
+      }
+    );
+    const heapSpaceNameAttributeName = `${this.namePrefix}.${V8_HEAP_SIZE_NAME_ATTRIBUTE}`;
+
+    meter.addBatchObservableCallback(
+      observableResult => {
+        if (this._scrapeQueue.length === 0) return;
+
+        const data = this._scrapeQueue.shift();
+        if (data === undefined) return;
+        for (const space of data) {
+          const spaceName = space.space_name;
+
+          observableResult.observe(heapLimit, space.space_size, {
+            [heapSpaceNameAttributeName]: spaceName,
+          });
+
+          observableResult.observe(heapSpaceUsed, space.space_used_size, {
+            [heapSpaceNameAttributeName]: spaceName,
+          });
+
+          observableResult.observe(
+            heapSpaceAvailable,
+            space.space_available_size,
+            {
+              [heapSpaceNameAttributeName]: spaceName,
+            }
+          );
+
+          observableResult.observe(
+            heapSpacePhysical,
+            space.physical_space_size,
+            {
+              [heapSpaceNameAttributeName]: spaceName,
+            }
+          );
+        }
+      },
+      [heapLimit, heapSpaceUsed, heapSpaceAvailable, heapSpacePhysical]
+    );
+  }
+
+  internalEnable(): void {}
+
+  internalDisable(): void {}
+
+  protected scrape(): HeapSpaceInfo[] {
+    return v8.getHeapSpaceStatistics();
```

Comment:
same comment as previous, the scrape queue here is problematic and can be unaligned with the periodic exporting of metrics. so just eliminate it and export directly in the callbacks.